### DISCUSSION TITLE
Custom editor with WYSIWYG capability

### DISCUSSION THREAD

**Discussion by ronsig** [microsoft/vscode-discussions#634] [Extension Development QnA] [answered]
I'm starting my first major extension work.
We need a WYSIWYG custom editor for drag & dropping controls, resizing etc.

I was sure there'd be examples of that, but could find none (don't get me started with Bard/ChatGPT which give code samples that cannot be compiled). I can't use the Draw.io Integration sample except asa general reference, as it's going to be a commercial product.

So my basic question is - should my editor inherit from 'TextEditor'? implement 'CustomEditorProvider' (which file type should be in the template? implements CustomEditorProvider<XYZ> )? 
It's all confusing, and I searched and could not find reasonable documentation.

Thanks!
Ron

**Comment by r3m0t** [comment, marked as answer]
This example is MIT licensed- https://github.com/microsoft/vscode-hexeditor

https://code.visualstudio.com/api/extension-guides/custom-editors#customeditor-vs-customtexteditor

> 
When trying to decide which type of custom editor to use, the decision is usually simple: if you are working with a text based file format use CustomTextEditorProvider, for binary file formats use CustomEditorProvider

**Reply by ronsig** [reply]
Thanks, Tomer

I already came across the last sentence. It's still unclear. I have a text input file that I want the editor to use in order to display controls. E.g. Button on position 4,14, sized 20X30. So the input is textual, but the output is not. In addition, should I use CustomTextEditorProvider and extend it to process the input information? Or a CustomEditorProvider, even though I don't have any binary file. And in that case, what template should be used - CustomEditorProvider<???>

Thanks

**Reply by r3m0t** [reply]
Hi, yes it's about whether the file on disk (file input) is text. Follow this example https://github.com/microsoft/vscode-extension-samples/blob/main/custom-editor-sample/src/catScratchEditor.ts

* To read the file, use `TextDocument.getText`, or be efficient for very large files by using `TextDocument.lineAt/lineCount`
* To apply changes to the file when a user drags something, use `applyEdit`
* VS Code will show the "modified" indicator and the file will only get written when the user uses Save/Save All/Auto Save

To test, you can use the command "View: Reopen Editor With Text Editor", this shows the text contents of the document as VS Code sees it.

For generics, you can see more detail in the actual `vscode.d.ts` API description - https://github.com/microsoft/vscode/blob/1.78.2/src/vscode-dts/vscode.d.ts#L9153

**Comment by alefragnani** [comment]
Hi,

Based on your description, I think both should work. But to help with your decision, you could look at the sample provided in the VS Code Extension Samples repo https://github.com/microsoft/vscode-extension-samples/tree/main/custom-editor-sample. It shows you both scenarios.

Hope this helps

**Reply by ronsig** [reply]
Thanks, I checked both before posting. Was still left confused, especially regarding the type I should pass to  CustomEditorProvider<???>
